File: android/app/src/main/java/pl/kinowo/ui/common/CardSpacingStyle.kt
package pl.kinowo.ui.common

import androidx.compose.runtime.compositionLocalOf
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp

/**
 * Live-tunable vertical gaps in a [pl.kinowo.ui.list.FilmCard], read from
 * [LocalCardSpacingStyle] by the card and its showings block.
 *
 * The defaults are the SHIPPING gaps, dialled in on the tuning screen — so
 * production (which never provides its own value) renders exactly as before.
 * The non-prod `ShowtimeTuningScreen` injects an edited copy via
 * `CompositionLocalProvider` to preview spacing changes against the real card.
 * Each field is a DISTINCT spacing property already present in the card layout.
 * Mirrors the role of [ShowtimeChipStyle] for chip styling.
 */
data class CardSpacingStyle(
    /** Gap below the title, above the runtime/year meta pills. */
    val titleToMeta: Dp = 1.dp,
    /** Gap below the meta pills, above the rating badges. */
    val metaToRatings: Dp = 6.dp,
    /** Gap below the ratings, above the showings block. */
    val ratingsToShowings: Dp = 5.dp,
    /** Gap below a day label, above its first cinema name (or showtimes when the
     *  cinema header is hidden). Independent of [showingsBlock] so the day↔cinema
     *  distance can be dialled separately. 0 = cinema name sits directly under
     *  the day label. */
    val dayToCinema: Dp = 0.dp,
    /** Spacing for the rest of the showings block (cinema↔pills, pills↔next
     *  cinema, day-block↔day-block). The day↔cinema gap is [dayToCinema]. */
    val showingsBlock: Dp = 3.dp,
)

/** Spacing driving every film card. The default equals today's shipping gaps, so
 *  production renders exactly as before; the tuning screen overrides it through
 *  `CompositionLocalProvider`. */
val LocalCardSpacingStyle = compositionLocalOf { CardSpacingStyle() }

File: ios/Kinowo/Views/CardSpacingStyle.swift
import SwiftUI

/// Live-tunable vertical gaps inside a `FilmCardView`, read by the card and
/// `ShowingsView` from the environment.
///
/// Production never injects one, so the defaults below — which mirror the
/// hard-coded spacings the card used before this struct existed — reproduce the
/// shipping layout point for point. The non-prod `ShowtimeTuningScreen` swaps in
/// an edited copy to preview gap changes against the real `FilmCardView` before
/// we bake new values into the layout. Keep these defaults in lockstep with the
/// literals the card would otherwise hard-code.
///
/// We expose only the gaps that already exist as *distinct* spacing properties.
/// `sectionSpacing` is deliberately one lever for the whole title→meta→ratings
/// →showings stack (the card uses a single `VStack(spacing:)` there), matching
/// that shared spacing rather than splitting it.
struct CardSpacingStyle: Equatable {
    /// `FilmCardView`'s inner `VStack(spacing:)` — the shared gap between
    /// title, meta pills, ratings, and the showings block.
    var sectionSpacing: CGFloat = 8
    /// Extra clearance below the ratings flow (`RatingBadgesView.padding(.bottom)`),
    /// on top of `sectionSpacing`, so a wrapped last rating row can't crowd the
    /// showings block.
    var ratingsBottom: CGFloat = 14
    /// `ShowingsView`'s outer `VStack(spacing:)` — between day blocks, and
    /// between cinema blocks within a day.
    var showingsBlock: CGFloat = 6
    /// Gap below a day label, above its first cinema name / pill row.
    /// Independent of `showingsBlock` so the day↔cinema distance can be dialled
    /// separately (the Android twin is `CardSpacingStyle.dayToCinema`). Default
    /// equals the old shared `showingsBlock` gap, so production is unchanged.
    var dayToCinema: CGFloat = 6
    /// Padding above each day label.
    var dayLabelTop: CGFloat = 4
    /// `ShowingsView`'s per-cinema inner `VStack(spacing:)` — cinema label to
    /// its pill row.
    var cinemaToPills: CGFloat = 4
    /// `FlowLayout(lineSpacing:)` between wrapped rows of showtime pills.
    var pillRowSpacing: CGFloat = 4
}

private struct CardSpacingStyleKey: EnvironmentKey {
    static let defaultValue = CardSpacingStyle()
}

extension EnvironmentValues {
    var cardSpacingStyle: CardSpacingStyle {
        get { self[CardSpacingStyleKey.self] }
        set { self[CardSpacingStyleKey.self] = newValue }
    }
}
